<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Splicing QTL • leafcutter</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">leafcutter</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Installation.html">Installation</a>
</li>
<li>
  <a href="../articles/Usage.html">Differential splicing</a>
</li>
<li>
  <a href="../articles/Visualization.html">Visualization</a>
</li>
<li>
  <a href="../articles/sQTL.html">SplicingQTL</a>
</li>
<li>
  <a href="../reference/index.html">R functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/davidaknowles/leafcutter">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Splicing QTL</h1>
            
          </div>

    
    
<div class="contents">
<p>Splicing QTL analysis is a little more involved than differential splicing analysis, but we provide a script <code>scripts/prepare_phenotype_table.py</code> intended to make this process a little easier. We assume you will use <a href="http://fastqtl.sourceforge.net/">FastQTL</a> for the sQTL mapping itself, but reformatting the output if you want to use another tool (e.g. <a href="http://www.bios.unc.edu/research/genomic_software/Matrix_eQTL/">MatrixEQTL</a> ) should be reasonably straightforward. The script is pretty simple: a) calculate intron excision ratios b) filter out introns used in less than 40% of individuals or with almost no variation c) output these ratios as gzipped txt files along with a user-specified number of PCs.</p>
<p>You’ll need the <code>sklearn</code> and <code>scipy</code> Python packages installed, e.g.</p>
<pre><code>pip install sklearn</code></pre>
<p>for the PCA calculation.</p>
<p>Usage is e.g.</p>
<pre><code>python scripts/prepare_phenotype_table.py example_data/testYRIvsEU_perind.counts.gz -p 10</code></pre>
<p>where <code>-p 10</code> specifies you want to calculate 10 PCs for FastQTL to use as covariates.</p>
<p>FastQTL needs <code>tabix</code> indices. To generate these you’ll need <code>tabix</code> and <code>bgzip</code> which you may have as part of <code>samtools</code>, if not they’re now part of <code>htslib</code>, see <a href="https://github.com/samtools/htslib">https://github.com/samtools/htslib</a> for installation instructions (alternatively <code>apt-get install tabix</code> worked for me in Ubuntu 14.04). With these dependencies installed you can run the script created by and pointed to by the output of <code>prepare_phenotype_table.py</code>, e.g. <code>example_data/testYRIvsEU_perind.counts.gz_prepare.sh</code>.</p>
<p>We assume you’ll run FastQTL separately for each chromosome: the files you’ll need will have names like <code>testYRIvsEU_perind.counts.gz.qqnorm_chr21.gz</code>. The PC file will be e.g. testYRIvsEU_perind.counts.gz.PCs.</p>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Knowles, Yang Li, Jack Humphrey, Jonathan Pritchard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
