<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html>
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1.0">
<title>Leafcutter Shiny App • leafcutter</title>
<!-- jquery --><script src="https://code.jquery.com/jquery-3.1.0.min.js" integrity="sha384-nrOSfDHtoPMzJHjVTdCopGqIqeYETSXhZDFyniQ8ZHcVy08QesyHcnOUpMpqnmWq" crossorigin="anonymous"></script><!-- Bootstrap --><link href="https://maxcdn.bootstrapcdn.com/bootswatch/3.3.7/cerulean/bootstrap.min.css" rel="stylesheet" crossorigin="anonymous">
<script src="https://maxcdn.bootstrapcdn.com/bootstrap/3.3.7/js/bootstrap.min.js" integrity="sha384-Tc5IQib027qvyjSMfHjOMaLkfuWVxZxUPnCJA7l2mCWNIpG9mGCD8wGNIcPD7Txa" crossorigin="anonymous"></script><!-- Font Awesome icons --><link href="https://maxcdn.bootstrapcdn.com/font-awesome/4.6.3/css/font-awesome.min.css" rel="stylesheet" integrity="sha384-T8Gy5hrqNKT+hzMclPo118YTQO6cYprQmhrYwIiQ/3axmI1hQomh7Ud2hPOy8SP1" crossorigin="anonymous">
<!-- pkgdown --><link href="../pkgdown.css" rel="stylesheet">
<script src="../jquery.sticky-kit.min.js"></script><script src="../pkgdown.js"></script><!-- mathjax --><script src="https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <div class="container template-vignette">
      <header><div class="navbar navbar-default navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">leafcutter</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
<li>
  <a href="../articles/Installation.html">Installation</a>
</li>
<li>
  <a href="../articles/Usage.html">Differential splicing</a>
</li>
<li>
  <a href="../articles/Visualization.html">Visualization</a>
</li>
<li>
  <a href="../articles/sQTL.html">SplicingQTL</a>
</li>
<li>
  <a href="../reference/index.html">R functions</a>
</li>
      </ul>
<ul class="nav navbar-nav navbar-right">
<li>
  <a href="https://github.com/davidaknowles/leafcutter">
    <span class="fa fa-github"></span>
     
  </a>
</li>
      </ul>
</div>
<!--/.nav-collapse -->
  </div>
<!--/.container -->
</div>
<!--/.navbar -->

      
      </header><div class="row">
  <div class="col-md-9">
    <div class="page-header toc-ignore">
      <h1>Leafcutter Shiny App</h1>
            
          </div>

    
    
<div class="contents">
<p>Written by Jack Humphrey, David Knowles &amp; Yang Li</p>
<p>To see an example of the leafcutter shiny app in action without installing anything take a look at <a href="https://leafvis.shinyapps.io/leafvis/" class="uri">https://leafvis.shinyapps.io/leafvis/</a>. This shows leafcutter differential splicing results for a comparison of 10 brain vs. 10 heart samples (5 male, 5 female in each group) from <a href="https://www.gtexportal.org/home/">GTEx</a>.</p>
<p>The LeafCutter Shiny app has been tested on macOS 10.12 and Ubuntu 14.04.</p>
<div id="example" class="section level4">
<h4 class="hasAnchor">
<a href="#example" class="anchor"></a>Example</h4>
<p>If you’ve already installed leafcutter then you can easily fire up a local shiny app on the same GTEx example data. Navigate into the leafvis directory:</p>
<pre><code>cd leafvis</code></pre>
<p>and then download the example data using</p>
<pre><code>./download_example.sh</code></pre>
<p>This will download and unzip a folder called ‘example’. Now starting the shiny app should be as easy as</p>
<pre><code>./run_leafvis.R example/Brain_vs_Heart_results.Rdata</code></pre>
<p>You can even leave out the ‘example/Brain_vs_Heart_results.Rdata’ part as this is the default dataset ‘run_leafvis.R’ will try to use.</p>
</div>
<div id="annotation-code" class="section level4">
<h4 class="hasAnchor">
<a href="#annotation-code" class="anchor"></a>Annotation code</h4>
<p>The Shiny app includes functionality to label detected introns as annotated or cryptic. To do for a new dataset this an “annotation code” is needed, built from a “Gene Transfer Format” file appropriate for your genome. We provide pre-built annotation codes for hg19 and hg38 which can be downloaded by running</p>
<pre><code>./download_human_annotation_codes.sh</code></pre>
<p>from the ‘leafvis’ directory. To build a new annotation code you need:</p>
<ul>
<li><a href="http://bedtools.readthedocs.io/en/latest/">bedtools</a></li>
<li>a GTF transcript file of your choice - we recommend <a href="https://www.gencodegenes.org/">GENCODE</a>.</li>
<li>GNU Awk (gawk) &gt;= 3.1;</li>
</ul>
<p>On Mac you can install bedtools and GNU Awk using <a href="https://brew.sh/">homebrew</a>:</p>
<p><code>brew install gawk; brew install bedtools</code></p>
<p>Most Linuxes will have ‘gawk’ already. On Ubuntu ‘bedtools’ can be installed using</p>
<p><code>sudo apt-get install bedtools</code></p>
</div>
<div id="required-r-packages" class="section level4">
<h4 class="hasAnchor">
<a href="#required-r-packages" class="anchor"></a>Required R packages</h4>
<p>These should all have been installed when you installed LeafCutter:</p>
<ul>
<li>data.table</li>
<li>stringr</li>
<li>ggplot2</li>
<li>shiny</li>
<li>DT</li>
<li>intervals</li>
<li>reshape2</li>
<li>gridExtra</li>
<li>foreach</li>
<li>shinycssloaders</li>
</ul>
</div>
<div id="step-0--generate-the-annotation-files" class="section level3">
<h3 class="hasAnchor">
<a href="#step-0--generate-the-annotation-files" class="anchor"></a>Step 0. Generate the annotation files</h3>
<p>This step processes a given GTF to generate lists of exons, introns and splice sites. This step only has to be run once to use a particular GTF file with the app.</p>
<p><strong>Warning</strong>: this has only been tested with the GENCODE V26 human GRCh37/GRCh38 and mouse GRCm38.p5 GTF files.</p>
<pre><code>sh wrangleAnnotation.sh &lt;GTF&gt; &lt;annotation_code&gt;</code></pre>
<p><strong>GTF</strong> - a gene transfer format file, provided by GENCODE, Ensembl, etc. <strong>annotation_code</strong> - the path or folder name with a base name, eg <code>/path/to/gencode_hg38</code></p>
<p>This will create:</p>
<pre class="&lt;annotation_code&gt;_all_introns.bed.gz"><code>&lt;annotation_code&gt;_threeprime.bed.gz
&lt;annotation_code&gt;_fiveprime.bed.gz
&lt;annotation_code&gt;_all_exons.txt.gz</code></pre>
</div>
<div id="step-1--prepare-the-leafcutter-differential-splicing-results-for-visualisation" class="section level3">
<h3 class="hasAnchor">
<a href="#step-1--prepare-the-leafcutter-differential-splicing-results-for-visualisation" class="anchor"></a>Step 1. Prepare the LeafCutter differential splicing results for visualisation</h3>
<p>This step annotates each intron in each cluster at a given false discovery rate.</p>
<pre><code>Rscript prepare_results.R --iFolder &lt;iFolder&gt; \
                         --oFolder &lt;oFolder&gt; \
                         --support &lt;support&gt; \
                         --annotation_code &lt;annotation_code&gt; \
                         --code &lt;code&gt; \
                         --FDR &lt;FDR&gt; \                       </code></pre>
<p><strong>iFolder</strong> - the folder that contains the results of the differential intron excision analysis. The script assumes all results share the same path and <strong>code</strong>.</p>
<p><strong>oFolder</strong> - the folder where the results will be written.</p>
<p><strong>support</strong> - the same support file as used for the previous analysis.</p>
<p><strong>annotation_code</strong> - as before.</p>
<p><strong>code</strong> - the same code used for the rest of the analysis, eg <code>testYRIvsEU</code></p>
<p><strong>FDR</strong> - the benjamini-hochberg false discovery rate with which to filter the results.</p>
<p>This will create the Rdata object <code>&lt;oFolder&gt;/results/&lt;code&gt;_results.Rdata</code>. The file ‘prepare_example.sh’ shows how this would be done for the example dataset if you wanted to rebuild ‘Brain_vs_Heart_results.Rdata’.</p>
</div>
<div id="step-2--visualise-the-results" class="section level3">
<h3 class="hasAnchor">
<a href="#step-2--visualise-the-results" class="anchor"></a>Step 2. Visualise the results</h3>
<pre><code>cd leafvis/
Rscript run_leafvis.R &lt;oFolder&gt;/results/&lt;code&gt;_results.Rdata</code></pre>
<p>This will load in the Rdata object and run the LeafCutter Visualisation App in your browser.</p>
</div>
<div id="features" class="section level3">
<h3 class="hasAnchor">
<a href="#features" class="anchor"></a>Features</h3>
<ul>
<li>Visualisation of significantly changed clusters</li>
</ul>
<div class="figure">
<img src="cluster_plot_example.png" alt="cluster plot example"><p class="caption">cluster plot example</p>
</div>
<ul>
<li>Visualisation of all clusters discovered in a gene</li>
</ul>
<div class="figure">
<img src="gene_plot_example.png" alt="gene plot example"><p class="caption">gene plot example</p>
</div>
<ul>
<li><p>PCA plot based on splice junction counts</p></li>
<li><p>All plots can be exported as PDF</p></li>
</ul>
</div>
</div>
  </div>

  <div class="col-md-3 hidden-xs hidden-sm" id="sidebar">
      </div>

</div>


      <footer><div class="copyright">
  <p>Developed by David Knowles, Yang Li, Jack Humphrey, Jonathan Pritchard.</p>
</div>

<div class="pkgdown">
  <p>Site built with <a href="http://hadley.github.io/pkgdown/">pkgdown</a>.</p>
</div>

      </footer>
</div>

  </body>
</html>
